### PR TITLE
Semantic Snippets - fix tabstop location for cursor in blocks

### PR DESCRIPTION
**Pull request by akhera99**
Gets the indentation of the line containing the opening brace and finds the indentation for the next line. Inserts that indentation as prepended leading trivia to the closing brace.

### CODE REVIEW

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs [hunk 1]
```diff
@@ -38,7 +31,49 @@ protected override void GetIfStatementConditionAndCursorPosition(SyntaxNode node
         {
             var ifStatement = (IfStatementSyntax)node;
             condition = ifStatement.Condition;
-            cursorPositionNode = ifStatement.Statement.SpanStart + 1;
+            var blockStatement = (BlockSyntax)ifStatement.Statement;
+
+            var triviaSpan = blockStatement.CloseBraceToken.LeadingTrivia.Span;
+
+            // Getting the midpoint location of the trivia to find where the cursor shoudl end up
+            cursorPositionNode = triviaSpan.Start + (triviaSpan.Length / 2) + 1;
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:39]: `cursorPositionNode` is a misleading name here. 

Also, why is it `/2` here? What makes the mid point special?

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs [hunk 2]
```diff
@@ -38,7 +31,49 @@ protected override void GetIfStatementConditionAndCursorPosition(SyntaxNode node
         {
             var ifStatement = (IfStatementSyntax)node;
             condition = ifStatement.Condition;
-            cursorPositionNode = ifStatement.Statement.SpanStart + 1;
+            var blockStatement = (BlockSyntax)ifStatement.Statement;
+
+            var triviaSpan = blockStatement.CloseBraceToken.LeadingTrivia.Span;
+
+            // Getting the midpoint location of the trivia to find where the cursor shoudl end up
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:38]:
```suggestion
            // Getting the midpoint location of the trivia to find where the cursor should end up
```

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs [hunk 3]
```diff
@@ -38,7 +31,49 @@ protected override void GetIfStatementConditionAndCursorPosition(SyntaxNode node
...
+            var parsedDocument = ParsedDocument.CreateSynchronously(document, cancellationToken);
+            var ifStatementSyntax = (IfStatementSyntax)node;
+            var openBraceLine = parsedDocument.Text.Lines.GetLineFromPosition(ifStatementSyntax.Statement.SpanStart).LineNumber;
+
+            var indentationOptions = new IndentationOptions(syntaxFormattingOptions);
+            var newLine = indentationOptions.FormattingOptions.NewLine;
+
+            var indentationService = parsedDocument.LanguageServices.GetRequiredService<IIndentationService>();
+            var indentation = indentationService.GetIndentation(parsedDocument, openBraceLine + 1, indentationOptions, cancellationToken);
+            var newIndentation = new IndentationResult(indentation.BasePosition, indentation.Offset + syntaxFormattingOptions.TabSize);
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:53]: worth commenting why the `Offset + TabSize` is correct here

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs [hunk 4]
```diff
@@ -38,7 +31,49 @@ protected override void GetIfStatementConditionAndCursorPosition(SyntaxNode node
...
+            return newIndentation.GetIndentationString(parsedDocument.Text, syntaxFormattingOptions.UseTabs, syntaxFormattingOptions.TabSize) + newLine;
+        }
+
+        protected override async Task<Document> AddIndentationToDocumentAsync(Document document, int position, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nearestStatement = FindAddedSnippetSyntaxNode(root, position, syntaxFacts);
+
+            if (nearestStatement is null)
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:62]: Do we ever expect this to be null?

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs (hunk 1 above)
Comment (akhera99) [reply src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs]: elaborated in the comment

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs (hunk 4 above)
Comment (akhera99) [reply src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:64]: Ideally it shouldn't, but if something goes very wrong when trying to find the snippet we return null

Comment (ryzngard) [reply src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:70]: IMO we should throw on that case, and have `RoslynDebug.AssertNotNull` to get NRT correct

File: src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs [hunk 5]
```diff
@@ -205,5 +213,10 @@ private async Task<Document> AddFormatAnnotationAsync(Document document, int pos
             document = document.WithSyntaxRoot(annotatedSnippetRoot);
             return document;
         }
+
+        protected virtual async Task<Document> AddIndentationToDocumentAsync(Document document, int position, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
```
Comment (ryzngard) [inline src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs:217]: Add some xml docs on why this exists, when it's called, and how this relates to formatting

**Review by ryzngard — APPROVED** [review]

File: src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs (hunk 4 above)
Comment (akhera99) [reply src/Features/CSharp/Portable/Snippets/CSharpIfSnippetProvider.cs:70]: I agree

**Pull request merged**